api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt

```diff
@@ -52,6 +55,38 @@ class BackendService(
             responseEntity.body.toMono().orNotFound { "Authentication body is empty" }
         }
 
+    /**
+     * Find current user [SaveUserDetails] by [authentication].
+     *
+     * @param authentication current user [Authentication]
+     * @return current user [SaveUserDetails]
+     */
+    fun findByAuthentication(authentication: Authentication): Mono<SaveUserDetails> = when (authentication) {
+        is UsernamePasswordAuthenticationToken -> findByName(authentication.name)
+        is OAuth2AuthenticationToken -> {
+            val source = authentication.authorizedClientRegistrationId
+            val nameInSource = authentication.name
+            findByOriginalLogin(source, nameInSource)
+        }
+        else -> Mono.empty()
+    }
+
+    /**
+     * Find current username by [authentication].
+     *
+     * @param authentication current user [Authentication]
+     * @return current username
+     */
+    fun findNameByAuthentication(authentication: Authentication?): Mono<String> = when (authentication) {
+        is UsernamePasswordAuthenticationToken -> authentication.name.toMono()
+        is OAuth2AuthenticationToken -> {
+            val source = authentication.authorizedClientRegistrationId
+            val nameInSource = authentication.name
+            findByOriginalLogin(source, nameInSource).map { it.name }
```

Comment:
and in this method just do commonMethod.name() ...

Reply:
When `authentication` is `OAuth2AuthenticationToken`, we must send a request, while when `authentication` is `UsernamePasswordAuthenticationToken`, there are to different variants.

Reply:
ok, then
```kotlin
fun findNameByAuthentication(authentication: Authentication?): Mono<String> = when (authentication) {
        is UsernamePasswordAuthenticationToken -> authentication.name.toMono()
        else -> findByAuthentication(authentication).map { it.name }
    }
```

just to minimize duplication

Reply:
Well, this one is ok, but there are two type casts...